### PR TITLE
fix(sdk): Allow uint8 images to be logged as wandb.Image()

### PR DESCRIPTION
**Pull request by nate-wandb**
Fixes
-----
- Fixes [WB-11242](https://wandb.atlassian.net/browse/WB-11242)

Description
-----------
Images loaded with torchvision.io.read_image would previously crash when logged as a wandb.Image() do to being normalized. This fix converts torch.uint8 images to `float` prior to logging them to solve this. 

Testing
-------
Added test_log_uint8_image() to test_data_types.py. Also passed all other tests in test_data_types.py as well and should only impact images of type torch.uint8. We also logged the randomly created image to wandb.ai with our branch and confirmed that the image created matches what is expected. 

### <samp>🤖 Generated by Copilot at 24f9ac6</samp>

> _Oh we're the wandb crew and we log what we do_
> _With tensors and images of every hue_
> _But when we find a bug in the `wandb.Image` class_
> _We heave away and fix it with a torch and a cast_

[WB-11242]: https://wandb.atlassian.net/browse/WB-11242?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/pytest_tests/unit_tests/test_data_types.py [hunk 1]
```diff
@@ -1525,3 +1525,21 @@ def test_numpy_arrays_to_list():
     assert conv(np.array((1, 2))) == [1, 2]
     assert conv([np.array((1, 2))]) == [[1, 2]]
     assert conv(np.array(({"a": [np.array((1, 2))]}, 3))) == [{"a": [[1, 2]]}, 3]
+
+def test_log_uint8_image():
+    # Create and save image
+    imarray = np.random.rand(100,100,3) * 255
```
Comment (dmitryduev) [inline tests/pytest_tests/unit_tests/test_data_types.py:1531]: Looks like the pre-commit hooks are not working for some reason... Could you try re-running `pre-commit install`?
separately, running `tox -e ruff-fix` should fix the linting issues here and elsewhere.

File: wandb/sdk/data_types/image.py [hunk 2]
```diff
@@ -287,6 +287,8 @@ def _initialize_from_data(
             )
             if hasattr(data, "requires_grad") and data.requires_grad:
                 data = data.detach()  # type: ignore
+            if str(data.dtype) == "torch.uint8":
```
Comment (dmitryduev) [inline wandb/sdk/data_types/image.py:290]: let's add a bit of paranoia here :)
```suggestion
            if hasattr(data, "dtype") and str(data.dtype) == "torch.uint8":
```

**Review by dmitryduev — APPROVED** [review]
🎉 nice, @nate-wandb @MBakirWB!! please address the minor comments, and this can go in.
also, feel free to add your future PRs to a future (target) release.

File: tests/pytest_tests/unit_tests/test_torch.py [hunk 3]
```diff
@@ -105,3 +105,11 @@ def test_watch_bad_argument(mock_run):
         ValueError, match="log must be one of 'gradients', 'parameters', 'all', or None"
     ):
         run.watch(net, log="bad_argument")
+
+def test_log_uint8_image(mock_run, capsys):
+    run = mock_run(use_magic_mock=True)
+    uint8_tensor = torch.rand([100, 100]).to(torch.uint8)
+    im = wandb.Image(uint8_tensor)
+    run.log({'uint8_image': im})
```
Comment (dmitryduev) [inline tests/pytest_tests/unit_tests/test_torch.py:113]:
```suggestion
    run.log({"uint8_image": im})
```

File: tests/pytest_tests/unit_tests/test_torch.py [hunk 4]
```diff
@@ -105,3 +105,11 @@ def test_watch_bad_argument(mock_run):
         ValueError, match="log must be one of 'gradients', 'parameters', 'all', or None"
     ):
         run.watch(net, log="bad_argument")
+
+def test_log_uint8_image(mock_run, capsys):
```
Comment (dmitryduev) [inline tests/pytest_tests/unit_tests/test_torch.py:109]: you don't seem to be using it here.
```suggestion
def test_log_uint8_image(mock_run):
```

File: tests/pytest_tests/unit_tests/test_torch.py [hunk 5]
```diff
@@ -105,3 +105,11 @@ def test_watch_bad_argument(mock_run):
...
+    run = mock_run(use_magic_mock=True)
+    uint8_tensor = torch.rand([100, 100]).to(torch.uint8)
+    im = wandb.Image(uint8_tensor)
```
Comment (dmitryduev) [inline tests/pytest_tests/unit_tests/test_torch.py:112]: looks like you're just trying to see if this fails, no? so that there's no need to try and log it?

**Comment by dmitryduev** [comment]
Looks like gh is having a hard time triggering circleci, will have to try again once it's working. still, need to fix the linting: https://app.circleci.com/pipelines/github/wandb/wandb/25263/workflows/e058f638-99b4-4d57-af64-80b5c75d3631/jobs/704770

File: tests/pytest_tests/unit_tests/test_data_types.py (hunk 1 above)
Comment (nate-wandb) [reply tests/pytest_tests/unit_tests/test_data_types.py:1531]: @dmitryduev ok, tried this and it says successful. Not sure we totally understand what this did though. Did it remove the blank line on 18?

Comment (dmitryduev) [reply tests/pytest_tests/unit_tests/test_data_types.py:1531]: you need to commit the linting results from ruff-fix -- note that code-check is still red in the ci.

File: tests/pytest_tests/unit_tests/test_torch.py (hunk 4 above)
Comment (nate-wandb) [reply tests/pytest_tests/unit_tests/test_torch.py:109]: Sorry, we actually didn't mean to leave this test in at all. We've removed this test entirely and just have the test that lives in test_data_types.py

File: tests/pytest_tests/unit_tests/test_torch.py (hunk 5 above)
Comment (nate-wandb) [reply tests/pytest_tests/unit_tests/test_torch.py:112]: Correct, we didn't mean to leave this test in. We'll remove the entire function

File: tests/pytest_tests/unit_tests/test_data_types.py (hunk 1 above)
Comment (nate-wandb) [reply tests/pytest_tests/unit_tests/test_data_types.py]: @dmitryduev we somehow broke `tox -e ruff-fix`  the second time we tried it so we just ran the linters that were failing in CI manually 🙃 We got it to work this way but might need some help doing this the right way next time

File: wandb/sdk/data_types/image.py (hunk 2 above)
Comment (nate-wandb) [reply wandb/sdk/data_types/image.py]: Updated 👍  thank you

File: tests/pytest_tests/unit_tests/test_torch.py (hunk 4 above)
Comment (nate-wandb) [reply tests/pytest_tests/unit_tests/test_torch.py]: @dmitryduev if you are ok without this check and just using our other test I'll mark this as resolved and merge now that we passed CI tests

**Review by dmitryduev — APPROVED** [review]
lgtm. i fixed up the test to use a temp file and a proper way to compare two np arrays
